doc/read-the-docs-site/adr/0005-inlining-saturated-functions.md

```diff
@@ -0,0 +1,206 @@
+
+For example:
+
+```haskell
+let f x y = 1
+    y = f q
+in f a b 
+```
+
+we should inline `f` on line 3 (even though it’s used more than once), because it’s saturated,
+i.e. it has all its arguments, and we can reduce it.
+([PLT-1146](https://input-output.atlassian.net/browse/PLT-1146))
+
+To do this we need to do some arity analysis of functions,
+so we know how many term/type arguments they have.
+We also need to decide whether or not to inline based on the call site, which we don’t currently.
+
+This ADR records the details of the design decisions we made for the arity analysis of functions.
+It also lists the weaknesses of the approach and plans to mitigate them.
+The mechanics of whether or not to inline based on the call site will be covered in a separate ADR.
+
+## Decision
+
+For a first round of implementation, we are adding a simplified pass.
+We will refine the implementation over time.
+In each round, we benchmark and test and investigate the results.
+
+### The rationale: Counting lambdas
+
+**Arity** is relevant when we talk about a fully applied function.
+A fully applied function has an arity of 0.
+The arity of a function is the number of arguments it takes minus the number of arguments applied.
+E.g., `\x.\y.x+y` has an arity of 2. `(\x.\y.x+y) 1` has an arity of 1.
+
+In this inline pass we check for a binding that is a function and is fully applied in the body.
+
+E.g.,
+
+```haskell
+let v = rhs in body
+```
+
+We pattern match the _rhs_ with `LamAbs` or `TyAbs` (lambda abstraction for terms or types),
+and count the number of lambdas.
+Then, in the _body_, we check for term or type application (`Apply` or `TyInst`)of _v_.
+
+If _v_ is fully applied in the body, i.e., if
+
+1. the number of type lambdas equals the number of type arguments applied, and
+2. the number of term lambdas equals the number of term arguments applied,
+
+we inline _v_, i.e., replace its occurrence with _rhs_ in the _body_ (**if other call site inlining
+conditions are satisfied**. We currently inline too little, this will be improved later, see below).
+For the rest of the discussion here we focus on the fully applied condition only.
+
+E.g., for
+
+```haskell
+let f = \x.\y.x+y in
+    f 2 4
+```
+
+the _body_ `f 2 4` is a function application of `f`. Since `f` is fully applied, we inline `f`.
+
+### Cases we miss
+
+The above case is straight forward. But there are other cases that
+may not give us the desired result if we just count the lambdas. Below are some examples.
+
+#### Example 1: function in body
+
+```haskell
+let f = \x . x
+     g = \y . f
+in g a
+```
+
+`f` and `g` each has an arity of 1. However, `g`'s _body_ includes `f` which also has an arity of 1.
+Since we only count the number of lambdas, `g` is fully applied, and we inline.
+But `g a` reduces to `f`, which has an arity of 1.
+So inlining `g` doesn't give the desired result of a fully reducible term in this case.
```

Comment:
I don't think this example is correct. We _should_ inline `g`, because inlining `g` produces `let f = \x . x in f`, which is better than the original.

Reply:
I was gonna discuss the other call site inlining condition too and with `sizeIsAcceptable`, `g` will not be inlined anyway. (We reject `LamAbs`) Perhaps I will just work on the code as is planned (counting lambdas, `acceptable` condition) and see how it goes. I can come back with results in the ADR. Bottom line is that this needs more work.

Reply:
> with sizeIsAcceptable, g will not be inlined anyway.

`g` _will_ be inlined, because once `g` is fully applied, its body is just a single variable, `f`.